src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/tracerecorder/TraceCollectingEventTracker.kt

```diff
@@ -0,0 +1,904 @@
+     * All this data replicate multiple arrays of [ManagedStrategy] indexed by thread id,
+     * but this class is intended to be stored in concurrent hash map keyed by [Thread].
+     *
+     * TODO: Unify this storage with [ManagedStrategy] and unify all code to track
+     *       method calls.
+     */
+    private class ThreadData(
+        /**
+         * ID of thread to use in trace, simply an increasing number.
+         */
+        val id: Int,
+        /**
+         * Collector to store all [TracePoint]s generated for this thread.
+         */
+        val collector: TraceCollector = TraceCollector(),
+        /**
+         * Current "tracked" stacktrace.
+         */
+        val stackTrace: MutableList<CallStackTraceElement> = arrayListOf(),
+        /**
+         * Additional stack information to resolve owner information for fields
+         * TODO: Unify with stackTrace
+         */
+        val shadowStack: MutableList<ShadowStackFrame> = arrayListOf(),
+        /**
+         * Tracks content of constants (i.e., static final fields).
+         * Stores a map `object -> fieldName`,
+         * mapping an object to a constant name referencing this object.
+         */
+        val constants: IdentityHashMap<Any, String> = IdentityHashMap<Any, String>(),
+        /**
+         * Stack of different analysis sections to filter-out unneeded events
+         */
+        val analysisSectionStack: MutableList<AnalysisSectionType> = arrayListOf(),
+        /**
+         * Last read constant name (i.e., static final field).
+         * We store it as we initialize read value after the trace point is created,
+         * so we have to store the trace point somewhere to get it later.
+         */
+        var lastReadConstantName: String? = null,
+        /**
+         * Last read trace point, occurred in the current thread.
+         * We store it as we initialize read value after the point is created,
+         * so we have to store the trace point somewhere to get it later.
+         */
+        var lastReadTracePoint: ReadTracePoint? = null
+    )
+
+    // We don't want to re-create this object each time we need it
+    private val analysisProfile: AnalysisProfile = AnalysisProfile(false)
+
+    // We don't use [ThreadDescriptor.eventTrackerData] because we need to list all descriptors in the end
+    private val threads = ConcurrentHashMap<Thread, ThreadData>()
+    // ID generator for [TracePoint]s
+    private val currentTracePointId = atomic(0)
+    // ID generator for [StackTraceElement]s
+    private val currentCallId = atomic(0)
+
+    override fun beforeThreadFork(thread: Thread, descriptor: ThreadDescriptor) = runInsideIgnoredSection {
+        // Don't init new threads forked from initial one if it is not enabled
+        if (TRACE_ONLY_ONE_THREAD) {
+            return
+        }
+
+        val td = threads[Thread.currentThread()] ?: return
```

Comment:
Let's use full names for the variables, for consistency with the rest of the code, e.g., `threadData` in this case.